### DISCUSSION TITLE
Keystrokes in commit dialog are a problem

### DISCUSSION THREAD

**Discussion by Spharad** [gitextensions/gitextensions#11873] [Q&A]
### Environment

- GitExtensions version: 3.5.4
- GIT version: 2.42.0.windows.2
- OS version: Windows NT 10
- .NET version: 4.8

### Issue description

I "accidentally" reset changes to a staged file. When I say accidentally, I thought I was typing in the comments section, and I typed "r e" which reset the current file.  This is a terrible thing.
This is not really a bug, since it's working as poorly as it was designed.  I hope you guys fix this because I have ran into this several times.  I'm so old that I've been developing since before the mouse, and I prefer the keyboard.
My suggestion is no NOT allow keystrokes on this dialog, and perhaps others, at least those that can have severe consequences.

### Steps to reproduce

1) Make some very important changes to several files, preferably over a few days so you have time to forget the important changes you made.  It also helps to be on a tight deadline and/or be sleep deprived.
2) Start typing a detailed commit message.
3) Click on one of the files to check what you did.
4) Forget to click back on the commit message window.  
5) Start typing any word that starts with "re".  R brings up the dialog:
![image](https://github.com/user-attachments/assets/ea78e8d4-842b-4019-8202-cddf87b00b95)
E clicks "Reset".
6) Congratu-frikken-lations, now you get to try to replicate what you did.  You probably get to guess which file was removed.  If you have the presence of mind to go to visual studio, and click undo, maybe you can recover what you did.

### Did this work in previous version of GitExtensions?

Who knows

**Comment by pmiossec** [comment]
Not a solution to your problem but you could have a look at that to recover reset staged files: https://stackoverflow.com/questions/11094968/in-git-how-can-i-recover-a-staged-file-that-was-reverted-prior-to-committing/58853981#58853981

**Reply by stenaught** [reply]
Yes, I did try that, but I had not staged my files.

**Comment by RussKie** [comment]
Your version is very old. I have a vague recollection of a similar bug, and
I'm fairly certain that it was fixed.
Please update

**Comment by RSchwoerer** [comment]
This still can occur with 4.2. 
It is a very interesting UX problem since you really want to have the keyboard shortcuts, I use them daily, and yet the accidental reset is a destructive behavior.
The reset dialog already defaults to cancel, so hitting `space` or `enter` will cancel. It is only the specific keystroke `r e` that causes the issue.


**steps to repro**
1. create pending changes
2. `ctrl`+`space` for commit dialog
3. `r e`
4. **all changes reset**


Edit to add, I tried the other popular, mollusc based, git gui and I could not find keyboard shortcuts for many operations, and certainly not for reset/discard, either the current file or all files. 
Might be worth considering removing at least the `e` keyboard accelerator from that dialog.

 Edit2 to add, if you just have one modified file, hitting `s` stages and jumps focus to the commit comment box. Maybe the commit comment box should be the initial focus. ~~That does prevent really any keyboard shortcuts however, as that box is focused it is not possible (that I found) to get out with just the keyboard~~. Wrong, `alt`+`whatever` shortcuts work.

**Comment by RussKie** [comment]
> Maybe the commit comment box should be the initial focus.

That'd make sense if all files were staged...



> * create pending changes
> * `ctrl`+`space` for commit dialog
> * `r e`

Admittedly, I generally use mouse and don't rely on keyboard as much (unlike some other users) but I'm somewhat confused - why would you start typing as the dialog opens?

. . .

The mnemonics (and the existing behaviour) have been like this for many years now. We had a related issue several years ago but other than that we didn't have any complaints (and there are several hundred thousand of users). I'm not disputing that the experience of losing changes is horrible (and I'm sorry for that), but there's a material risk of annoying a significant number of users by changing the existing and well-established behaviours.

**Comment by mstv** [comment]
As a workaround, you could remove the mnemonic from the translation file (`*.xlf`), i.e. remove `"&amp;"` from the last line in this section:
```
  <file datatype="plaintext" original="FormResetChanges" source-language="en">
    <body>
      <trans-unit id="$this.Text">
        <source>Reset changes</source>
        <target />
      </trans-unit>
      <trans-unit id="btnCancel.Text">
        <source>&amp;Cancel</source>
        <target />
      </trans-unit>
      <trans-unit id="btnReset.Text">
        <source>R&amp;eset</source>
```

**Reply by mstv** [reply]
Alternatively, you can change or remove the hotkey (as mentioned in https://github.com/gitextensions/gitextensions/issues/9552#issuecomment-910703672):

![image](https://github.com/user-attachments/assets/df209add-ec67-4220-9aac-7605a4f14196)

![image](https://github.com/user-attachments/assets/29c3a8ed-6e6e-4966-936f-ac8b414727c5)

**Reply by stenaught** [reply]
That will help me, but it's still a time bomb for those who are unaware of this issue.

**Comment by mstv** [comment]
duplicate of #9552, related to #8467
converting this into a discussion

**Reply by stenaught** [reply]
The duplicate was also moved to a discussion, and then closed.  This should be an issue, since the downside is potentially terrible.
